mache/discover.py

```diff
@@ -27,11 +29,26 @@ def discover_machine():
     elif hostname.startswith('cooley'):
         machine = 'cooley'
     elif hostname.startswith('cori'):
-        warnings.warn('defaulting to cori-haswell.  Use -m cori-knl if you'
-                      ' wish to run on KNL.')
+        warnings.warn('defaulting to cori-haswell.  Explicitly specify '
+                      'cori-knl as the machine if you wish to run on KNL.')
         machine = 'cori-haswell'
     elif hostname.startswith('gr-fe'):
         machine = 'grizzly'
-    else:
-        machine = None
+    elif 'NERSC_HOST' in os.environ:
+        hostname = os.environ['NERSC_HOST']
+        if hostname == 'perlmutter':
+            # perlmutter's hostname is too generic to detect, so relying on
+            # $NERSC_HOST
+
+            warnings.warn('defaulting to pm-cpu.  Explicitly specify '
+                          'pm-gpu as the machine if you wish to run on '
+                          'GPUs.')
+            machine = 'pm-cpu'
+        elif hostname == 'unknown':
+            raise ValueError(
+                'You appear to have $NERSC_HOST=unknown.  This typically '
+                'indicates that you \n'
+                'have an outdated .bash_profile.ext or similar.  Please '
+                'delete that file, \n'
```

Comment:
I have a lot of useful stuff in my `~/.bash_profile.ext`, so I wouldn't recommend just deleting that file no matter what. Perhaps replace "Please delete that file" with "Please edit that file or delete if necessary".